### ISSUE TITLE
Implement option to use BSP2 internal search algorithm at all times [Was: Just updated 05.02.21, bsp2 freezes, firefox not loading new tab]

### ISSUE THREAD

**Issue by ttggzz** [aaFn/Bookmark-search-plus-2#185]
Hello, I just updated, was using bsp2 yesterday no problems. Put a search term in I used yesterday and results showed a folder and several links. Clicked on the folder, nothing happened. Clicked on links also found in search, no go. Also, it seemed I could not open a new link in firefox. Internet is fine, another browser I have open loaded pages. I disabled bsp2 and was able to open new links in firefox after restarting. Posting this then I will try bsp2 again. thanks
  
edit: firefox 88 on Linux, same as used yesterday, nothing changed in firefox

edit: ok, I have an embarrassing number of tabs and windows open, not going to say :)  just tried bsp2 on a different profile with barely anything open and it's working normally, nice and snappy. Something must have reached a tipping point, I was already seeing overall lag in firefox. 

I will note bsp2 doesn't completely freeze, you can close it, just nothing happens when I click on a search result. And, no new tabs can be opened in firefox until you restart. Unfortunately I was just gearing up to use bsp2 to get this profile under control, d'oh

**Comment by aaFn** [comment]
Hello @ttggzz, ok, thank you for the update.
I just verified on Linux with Firefox 87 and 88, with more than 500 tabs open in 2 FF windows on one of my instances, and all looks fine on my side.

It looks like your problem is on opening new tabs, or on loading a new page in current tab.

Maybe 2 things to do when you have the problem:
1. Look at about:performance , and see if there is anything abnormal there, and also where is BSP2 if you sort by Energy impact, or by Memory.
2. Open the browser console (Ctrl+Shift+J), and look for any error message from BSP2 when you click on a link

Also, you could preventively clear the cache, that can't hurt from time to time .. (Edit -> Preferences -> Cookies and Site data)

Let me know what you find, aaFn.

**Comment by ttggzz** [comment]
Thank you much appreciated... I just tried something before getting your message. I clicked on one link in the search results, and walked away. 10 minutes later I see the link has opened below, showing its position in the library. This would be somewhat on a par with how firefox has been lagging with how much I have open, but only maybe 20 seconds or so, and intermitently.

Anyway, I was then able to open that link from bsp2. Also able to open links independently from firefox. I will have to get back to your suggestions later, just wanted to update on that before heading out. thanks

**Comment by aaFn** [comment]
Ok, seems that your Firefox threads are using up all their CPU from what you say ... about:performance should tell you what is hogging the CPU, can be a particular page, or an add-on (BSP2 or another), you'll see both types.

Could be interesting to also open the Windows task manager and see how much Firefox is consuming in % cpu.

8000 tabs ! Wow .. never got there 😄 ! (BSP2 is only using the browser.tabs API a few, when creating a new tab or loading in tab a clicked bookmark, and when you switch tabs to update the status of the star to show that we are on a bookmarked page, so I can't think that it could impact tab switching and tab operations by itself)

Let's see what you get ..

**Comment by ttggzz** [comment]
I cleared the cache which was 1 GB web content 432MB cookies/site data. In Task Manager I didn't see anything extreme, all the addons were showing. When I clicked to open a search term in BSP2 all the other addons disappeared in the Task Manager list. Then they came back after the search term loaded a few minutes later. 

I'm not sure what to look for in the browser console, I don't see bsp2 mentioned there in the error list. Thought I'd mention I use another tab-intensive addon, All Tabs Helper. I tried disabling it, didn't seem to make any difference. All Tabs Helper uses "container tabs", can't say I have a working knowledge of what that is, my eyes glaze over reading about them. I am going to try hacking away at closing some tabs for now.

**Comment by aaFn** [comment]
Ok, how many bookmarks do you have ? (in BSP2 options, you will see the stats)

And do you use tags / keywords in bookmarks ? I noticed the FF search API gets very slow and cpu hungry in such cases, when many bookmarks.
If so => Try using a different search option in the magnifier glass button, at left of BSP2 search bar = e.g. select the "Current folder tree" option, and start your next search by placing first your cursor at top of the bookmark tree (usually the personal toolbar).
Indeed, when a non default option is selected, I am using my own search algorithm in BSP2, not the FF API one. It should be much faster.

For the traces / browser console, sorry, I forgot to mention, when it is open, select the parameter wheel on ht top right, and activate "Show Content Messages" to see all add-ons messages.
BSP2 messages will then appear. No easy way to locate them, but they usually come from background.js or panel.js files.

If you confirm that you have plenty of bookmarks (which I suspect from your decription), and that the internal search algorithm is faster for you, I will add an option to bypass the FF API search to use only mine.

Thanks for all those details, aaFn.

**Comment by ttggzz** [comment]
Thanks, yeah I was just remembering about bookmarks, here's my stats:
Bookmarks:            76471
Favicons to fetch:    0 (0.0%)
Folders:              5013
Separators:           18443
Oddities:             0
If you mean something that I add myself when bookmarking, no I do not use tags/keywords. I just put things in folders that work for me, but only with bsp2 :) and earlier the old show/go parent folder add-on. 

I will try that other search option. I wanted to restate the oddity that the previous day bsp2 was working normally, albeit slight lag like everything else in that profile's bloated session content. Firefox update to 88 had been in place for several days of heavy use, my Linux system had been updated same, and no other addons were updated. Also, I doubled my ram from 16 to 32GB and had been using that ok. I just started working, did your update, then bsp2 fell down. And this was with many browser restarts, so at any given time there were only 30 or so tabs open of the 8,000

**Comment by aaFn** [comment]
Yes, you're doing it right.

Ok, great, so this is what I was suspecting that the FF API search is very long, because you have a lot of bookmarks, no surprise !

Not sure why it was more or less ok before the BSP2 update, and then sluggish / cpu consuming after that .. this is not a part I touched on 2.0.90 (supposing you were at 2.0.89 just before).
I tried on an instance where I have 30000 bookmarks and 10000 folders and where I had 2.0.89, then updated to 2.0.90, but did not get the effects you observed, all remains fast after update.

I will add an option in 2.0.91 to not use FF API search on default parameters, and to use the BSP2 internal algorithm instead, as a performance option (unchecked by default, so you will have to set it).
Meanwhile, if the problem persist on the default search, you can use the workaround, if not too cumbersome.

Let me know if you see anything else, thanks, aaFn.

**Comment by ttggzz** [comment]
Yes, I believe I would have been at 2.0.89, I manually check for updates frequently. Excellent, I will look for that update and in the meantime use the workaround. Thanks, and nice work! Relieved (very) that you had this possibility already built in. One of the great mysteries of our time: Firefox does not have this feature :)

There are a couple of things, one I believe is already in the preferences, I would like a bit more of a delay before the search I am typing activates. Or, now that I think about it, I'd actually prefer to fully type in my search then manually click it in with 'Enter'. 

Although in general I'd prefer using 'Enter' this request is related to some kind of system lag. If I make a mistake when typing or want to change something there is a lag of several seconds while the mistake loads before I can continue and change it. Maybe related to my library size (?), I can live with that lag if it's only happening after I manually activate the search. 

The other thing is also system lag, when I right click a link and hit 'i' to modify, it pauses then opens in a tiny window then after some seconds opens in the full view. This seems to vary widely, the lag. Also I will note that if I can see the tiny window I can begin modifying. 

Just testing now the modify window is opening pretty snappy, something must make the response vary. Perhaps because I've had bsp2 open for a bit? Quite often I have just opened bsp2 and am quickly saving a link with my own modifiers. 
thanks

**Comment by aaFn** [comment]
Correct, use the `Wait for "Enter" in search box to launch search (no trigger on key typing inactivity timeout)` option to do what you want on not having time out. It was made for people like you who have a lot of bookmarks, and long search times because of FF API.

On the second effect when you open the Property window, this is again a limitation from FF 😢 :
- I cannot open the window with the right position, due to a bug it seems in FF API
- So I have first to open it with very small size to make it as less conspicuous as possible, in its default position chosen by FF
- And then I have to move it to the right position, and enlarge it to right size at the same time
- Even worse, from past tests with other people having FF lags like you, it seems that some times FF does my first step in 2 smaller steps, one opening the window very very big (full screen size ?), and then making it very small ... ugly

In short, nothing I can do about for now = until I can create a window with right size and position in only one step, we all have to live with this workaround ...

**Comment by ttggzz** [comment]
OK, thanks, lookin' good! Painful to think how I suffered with that auto-fill search feature alterable by spending one minute in the preferences :) Thanks for the explanation on the Properties window, it helps to now at least understand what I'm seeing. 

I drag and drop when bookmarking, that works best for me with my trackball mouse habits. I notice sometimes when I enter the list to drop the lateral position of the tree would jerk around. I was going to ask about somehow locking that but after some observation I found I just needed to change my technique. 

I realized any hovering near the right hand border is what is causing the column to move. I have started to immediately hug the left hand side before I zoom in on where I want to make my drop, problem solved. 

I was having a few second thoughts about all those windows I closed. Not relishing the thought of trying to restore that behemoth mess of a session I stumbled on an interesting technique. In history I opened "older than 6 months", then scrolling through the list I was able to see various searches I had done, grouped loosely together historically. So far I've only done it for one topic but what I did was using control/click chose and reopened a manageable number (25) of choice looking related tabs from history, culled through those, then bookmarked. That actually seemed much easier than trying to sort through multiple windows with huge tab counts, which I was dreading.

**Comment by aaFn** [comment]
2.0.91 is now out with the option as promised above.
Let me know how that goes.

**Comment by ttggzz** [comment]
Ok, thanks, looking good... After choosing the new option in preferences (never use FF API) I also changed the option at the search bar back to "search all (default)" from "current folder tree" (had not changed that at first), after that change not sure what was going on as it froze up a couple times like before. I opened/closed bsp2 and firefox several times and now it is working smoothly. Nice work! Looking forward to trying the forthcoming "load bookmarks only when required."

**Comment by aaFn** [comment]
No, you do not need the cursor to be anywhere with the option.
It seems something else is going on.

I would suggest first you do a "Reload from FF API" in the options page, in case there is a de-sync between BSP2 and FF (with all these hangs, and stop / start, this is possible).
Of course, with your amount of bookmarks, expect the reload can take some time.

Then let's see if the freezes continue to happen.
If yes, then I would need an extract of the Browser console log = Ctrl+Shift+J, clean up its content with the trash icon, make sure content logging is enabled in the parameter wheel icon, then launch the operation which is freezing things ... wait it is complete .. and copy / paste here the total content of the console log

Thanks, aaFn.

**Comment by ttggzz** [comment]
"reload from Firefox API", that is the box in Performances, 'reload all bookmarks from Firefox now'? Should I have the 4 boxes in Performances unchecked/empty first?

**Comment by aaFn** [comment]
Yes, that button.

It is enabled when you do not tick on the `Revert to old behavior of loading full bookmark tree from Firefox API at each restart` option, which is the case by default, so I believe is your case.
No link with any other option.

**Comment by ttggzz** [comment]
when I click Reload from Firefox API, the button turns green for a fraction of a second, then goes to grayed out. Is that the expected behavior? Expecting it to maybe stay green for the time it took to reload, or something like that.

**Comment by aaFn** [comment]
Yes, when it turns gray, it means this is complete.
The sidebar is also reloaded when it is complete.
Let's see now if slow downs continue, and if yes, if you can get traces as said above.
Thanks, aaFn.

**Comment by ttggzz** [comment]
whoa... snappy! OK, I don't know exactly what I did before, but this time when I did "reload all bookmarks from FF now" I had bsp2 open. This time after the button went green for a split second a window opened in bsp2 indicating the reload was happening. It took 30-45 seconds (I did this again after restarting FF). I also had no choices ticked in "Performances". 

Then after the reload was done I chose Never use FF API in Performances, and made sure "Search all (default)" was chosen next to the search bar in bsp2. Off to the races!

I restarted FF several times, looks like it's working. Excellent, nice work! 

I tried the "only 1 open folder" in user interface. Not sure what I'm supposed to be seeing but if I click an item in the search window and then scroll up and down in the tree I can see the contents of all manner of folders. Are the various "show" options in the drop down next to the search bar where this type of control happens? 
thanks

**Comment by aaFn** [comment]
Good let's see if things remain stable then 🙂 

The option éonly 1 fodler" is for when you click on folders to open them.

The show function from search window does not change the state of folders to "Open" (junless you tick the corresponding option), and so it does not trigger the action to close other folders.
You can see that by noticing that the twistie of closed folders where the searched result is remains closed. That way, when you stop the search ("x" or ESC or empty the search box by hand), all content of such folders is hidden again, to restore the view before the search.

**Comment by ttggzz** [comment]
OK, still working. Hasn't been in heavy rotation but it "just works" every time, lookin' good! Regarding the folder info you provided, I will have to look at this more, haven't quite got my head around it due to not actually observing in action.

Question: if code or settings are changed in the native FF bookmark sidebar, will bsp2 then reflect those changes? Specifically, I searched last night on how to have all folders close when exiting the sidebar. This has annoyed me since forever, maybe I'm missing a setting somewhere, but reopening the sidebar can sometimes be like you didn't clean up last nights dinner or something :) 

Anyway, I found one promising result, marked "solved" in Reddit, only 2 months old! 
https://www.reddit.com/r/firefox/comments/m2pnrm/close_bookmark_tree_folders/
Looks like it is from our old friend alice0775 (go/show parent folder)
https://github.com/alice0775/userChrome.js/blob/master/78/AutoCloseBookMarkFolder_Fx37.uc.js

Sooo, never worked with this userChrome.js, if I figure out how to install it in FF will bsp2 then reflect that behavior as well? I'm thinking it might be nice to toggle such an option, maybe see how it goes first. 
thanks

**Comment by aaFn** [comment]
The open/closed state of folders in the FF native Bookmark Sidebar is not visible to add-ons, so BSP2 won't know.
BSP2 has its own open/closed state. In other words, view it as a display interface status, not a bookmark level status.

In BSP2, simply right-click on the Bookmarks Menu, then Advanced -> Collapse All in Branch.
That will do what you want (in BSP2 only).

**Comment by ttggzz** [comment]
Hello again, some freezing going on. Thanks for the tree/folder closing info, I'll have to look at that more later. Initial trying the "Collapse All in Branch" I would vote for a button next to the search bar. 

Although an initial freeze was earlier and happened with 30 or so tabs open (lots of youtube), a later one happened after I restarted browser then tried "Collapse all...". The browser console would not open. Tried that again and it froze, this time I had the browser console cleared and ready. Here is the contents. Unfamiliar with this, I have the FF info page on it open. Let me know if you need more info, thanks
`22:15:59.029 Error: Could not establish connection. Receiving end does not exist. 2 tabs.js:199:23
22:19:30.515 TypeError: tab is undefined background.js:1122:20
22:21:40.115 Some cookies are misusing the recommended “SameSite“ attribute 9
22:23:01.464 Firefox can’t establish a connection to the server at wss://alive.github.com/_sockets/u/33850550/ws?session=eyJ2IjoiVjMiLCJ1IjozMzg1MDU1MCwicyI6Njg5OTY3MjM0LCJjIjoxNzE2NDQ1MDk5LCJ0IjoxNjIxMDQzNTYxfQ%3D%3D--a15709267d5f16725adfc46ca5bc41b7d60329fd427c97b221ea4b21721190f0&shared=true. socket-worker-3f088aa2.js:1:2007
22:23:22.574 Firefox can’t establish a connection to the server at wss://alive.github.com/_sockets/u/33850550/ws?session=eyJ2IjoiVjMiLCJ1IjozMzg1MDU1MCwicyI6Njg5OTY3MjM0LCJjIjoxNzE2NDQ1MDk5LCJ0IjoxNjIxMDQzNTYxfQ%3D%3D--a15709267d5f16725adfc46ca5bc41b7d60329fd427c97b221ea4b21721190f0&shared=true. socket-worker-3f088aa2.js:1:2007
22:23:43.065 Firefox can’t establish a connection to the server at wss://alive.github.com/_sockets/u/33850550/ws?session=eyJ2IjoiVjMiLCJ1IjozMzg1MDU1MCwicyI6Njg5OTY3MjM0LCJjIjoxNzE2NDQ1MDk5LCJ0IjoxNjIxMDQzNTYxfQ%3D%3D--a15709267d5f16725adfc46ca5bc41b7d60329fd427c97b221ea4b21721190f0&shared=true. socket-worker-3f088aa2.js:1:2007
22:24:03.801 Firefox can’t establish a connection to the server at wss://alive.github.com/_sockets/u/33850550/ws?session=eyJ2IjoiVjMiLCJ1IjozMzg1MDU1MCwicyI6Njg5OTY3MjM0LCJjIjoxNzE2NDQ1MDk5LCJ0IjoxNjIxMDQzNTYxfQ%3D%3D--a15709267d5f16725adfc46ca5bc41b7d60329fd427c97b221ea4b21721190f0&shared=true. 2 socket-worker-3f088aa2.js:1:2007
22:24:32.359 Firefox can’t establish a connection to the server at wss://alive.github.com/_sockets/u/33850550/ws?session=eyJ2IjoiVjMiLCJ1IjozMzg1MDU1MCwicyI6Njg5OTY3MjM0LCJjIjoxNzE2NDQ1MDk5LCJ0IjoxNjIxMDQzNTYxfQ%3D%3D--a15709267d5f16725adfc46ca5bc41b7d60329fd427c97b221ea4b21721190f0&shared=true. 2 socket-worker-3f088aa2.js:1:2007
22:29:00.227 Error: Invalid tab ID: 2442 6 undefined
22:29:00.783 Unchecked lastError value: Error: Invalid tab ID: 2442 context-menu.js:40
22:29:00.783 Unchecked lastError value: Error: Invalid tab ID: 2442 context-menu.js:43
22:29:01.116 Unchecked lastError value: Error: Invalid tab ID: 2442 background.js:280
22:29:01.116 Unchecked lastError value: Error: Invalid tab ID: 2442 background.js:282
22:29:01.117 Unchecked lastError value: Error: Invalid tab ID: 2442 background.js:287
22:29:01.117 Unchecked lastError value: Error: Invalid tab ID: 2442 background.js:280
22:29:01.117 Unchecked lastError value: Error: Invalid tab ID: 2442 background.js:282
22:29:01.118 Unchecked lastError value: Error: Invalid tab ID: 2442 background.js:287
22:29:01.296 Unchecked lastError value: Error: Invalid tab ID: 2442 context-menu.js:40
22:29:01.296 Unchecked lastError value: Error: Invalid tab ID: 2442 context-menu.js:43
22:29:01.402 Unchecked lastError value: Error: Invalid tab ID: 2442 background.js:280
22:29:01.415 Unchecked lastError value: Error: Invalid tab ID: 2442 background.js:282
22:29:01.416 Unchecked lastError value: Error: Invalid tab ID: 2442 background.js:287
22:29:02.143 Error: Invalid tab ID: 2442 undefined
22:29:03.163 Error: Could not establish connection. Receiving end does not exist. 4 background.js:895:25
22:29:03.236 XHRGEThttps://r2---sn-ab5l6nzd.googlevideo.com/videoplayback?expire=1621030536&ei=KKKeYPbyK4eZhgbxqaTYBQ&ip=2601%3A188%3Ac302%3Acbc0%3A0%3A0%3A0%3A1236&id=o-AOl1jCqpxEwWGgbTvfMOgIUXVf4ro0u2aqhXAHIjR9aI&itag=251&source=youtube&requiressl=yes&mh=n3&mm=31%2C29&mn=sn-bvvbax-cvne%2Csn-ab5l6nzd&ms=au%2Crdu&mv=m&mvi=12&pl=34&initcwndbps=1778750&vprv=1&mime=audio%2Fwebm&ns=zg4iXPEq5BqKRfxq9gFOsGUF&gir=yes&clen=12640747&dur=766.181&lmt=1564046621512801&mt=1621008797&fvip=2&keepalive=yes&fexp=24001373%2C24007246&beids=23886210&c=WEB&txp=6301222&n=G7L2JZsf-_7lPQ&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRQIgEXaH0ACX_b41CTC3xzEaZ1qkzrbQXOL0DMQRt21f3LgCIQCvOCy_atbsGAYdcGc3NtIGbx4l2irm1wUzN9JqTWQAHQ%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRAIgN4DFcyhMtLFD8BIV-AEPkVGC2vlu_amx68AnSzZhYHECIFDvETEEY8fgXuhmHaCtPtz5EsJV_N46jsCNzyKVVH4v&alr=yes&cpn=xsPuLxxHGAbFOQSQ&cver=2.20210512.06.00&fallback_count=1&range=2241999-2422417&rn=31&rbuf=20091
[HTTP/1.1 403 Forbidden 231ms]

22:29:03.239 XHRGEThttps://r2---sn-ab5l6nzd.googlevideo.com/videoplayback?expire=1621030536&ei=KKKeYPbyK4eZhgbxqaTYBQ&ip=2601%3A188%3Ac302%3Acbc0%3A0%3A0%3A0%3A1236&id=o-AOl1jCqpxEwWGgbTvfMOgIUXVf4ro0u2aqhXAHIjR9aI&itag=243&aitags=133%2C134%2C135%2C136%2C137%2C160%2C242%2C243%2C244%2C247%2C248%2C278&source=youtube&requiressl=yes&mh=n3&mm=31%2C29&mn=sn-bvvbax-cvne%2Csn-ab5l6nzd&ms=au%2Crdu&mv=m&mvi=12&pl=34&initcwndbps=1778750&vprv=1&mime=video%2Fwebm&ns=zg4iXPEq5BqKRfxq9gFOsGUF&gir=yes&clen=33571236&dur=766.160&lmt=1564032433790164&mt=1621008797&fvip=2&keepalive=yes&fexp=24001373%2C24007246&beids=23886210&c=WEB&txp=6316222&n=G7L2JZsf-_7lPQ&sparams=expire%2Cei%2Cip%2Cid%2Caitags%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRAIgPWu00OXGyy17HfDQF-AiOU-kgEfqjZf8-5W6ch9cZ6cCIBuZP1i9ziCgEpSTiQSLL5Xqi2IppXTO8GfIDa8daEhZ&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRAIgN4DFcyhMtLFD8BIV-AEPkVGC2vlu_amx68AnSzZhYHECIFDvETEEY8fgXuhmHaCtPtz5EsJV_N46jsCNzyKVVH4v&alr=yes&cpn=xsPuLxxHGAbFOQSQ&cver=2.20210512.06.00&fallback_count=1&range=0-2837&rn=32&rbuf=0
[HTTP/1.1 403 Forbidden 217ms]

22:29:03.241 XHRGEThttps://r2---sn-ab5l6nzd.googlevideo.com/videoplayback?expire=1621030536&ei=KKKeYPbyK4eZhgbxqaTYBQ&ip=2601%3A188%3Ac302%3Acbc0%3A0%3A0%3A0%3A1236&id=o-AOl1jCqpxEwWGgbTvfMOgIUXVf4ro0u2aqhXAHIjR9aI&itag=250&source=youtube&requiressl=yes&mh=n3&mm=31%2C29&mn=sn-bvvbax-cvne%2Csn-ab5l6nzd&ms=au%2Crdu&mv=m&mvi=12&pl=34&initcwndbps=1778750&vprv=1&mime=audio%2Fwebm&ns=zg4iXPEq5BqKRfxq9gFOsGUF&gir=yes&clen=6052099&dur=766.181&lmt=1564046614080015&mt=1621008797&fvip=2&keepalive=yes&fexp=24001373%2C24007246&beids=23886210&c=WEB&txp=6301222&n=G7L2JZsf-_7lPQ&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRQIhAITMWFr4G01DPfKkUq6j5dl64t5qew6zdstZ1TstnG87AiBHPIHMF3a15Nk5jG-SgHSaSAqEMGkmYBIvSoLakIed-Q%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRAIgN4DFcyhMtLFD8BIV-AEPkVGC2vlu_amx68AnSzZhYHECIFDvETEEY8fgXuhmHaCtPtz5EsJV_N46jsCNzyKVVH4v&alr=yes&cpn=xsPuLxxHGAbFOQSQ&cver=2.20210512.06.00&fallback_count=1&range=0-1571&rn=33&rbuf=0
[HTTP/1.1 403 Forbidden 235ms]

22:29:03.243 XHRGEThttps://r2---sn-ab5l6nzd.googlevideo.com/videoplayback?expire=1621030536&ei=KKKeYPbyK4eZhgbxqaTYBQ&ip=2601%3A188%3Ac302%3Acbc0%3A0%3A0%3A0%3A1236&id=o-AOl1jCqpxEwWGgbTvfMOgIUXVf4ro0u2aqhXAHIjR9aI&itag=242&aitags=133%2C134%2C135%2C136%2C137%2C160%2C242%2C243%2C244%2C247%2C248%2C278&source=youtube&requiressl=yes&mh=n3&mm=31%2C29&mn=sn-bvvbax-cvne%2Csn-ab5l6nzd&ms=au%2Crdu&mv=m&mvi=12&pl=34&initcwndbps=1778750&vprv=1&mime=video%2Fwebm&ns=zg4iXPEq5BqKRfxq9gFOsGUF&gir=yes&clen=18204978&dur=766.160&lmt=1564032433766079&mt=1621008797&fvip=2&keepalive=yes&fexp=24001373%2C24007246&beids=23886210&c=WEB&txp=6316222&n=G7L2JZsf-_7lPQ&sparams=expire%2Cei%2Cip%2Cid%2Caitags%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRQIgGnOdWvOlWHTdWUzcrDLf29UuUK_u91MmLv-S64_ztEgCIQCCnU7BZ5uRGEg3TwzSnIeaEBrPBeZAS7gbQ55RG-dBtg%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRAIgN4DFcyhMtLFD8BIV-AEPkVGC2vlu_amx68AnSzZhYHECIFDvETEEY8fgXuhmHaCtPtz5EsJV_N46jsCNzyKVVH4v&alr=yes&cpn=xsPuLxxHGAbFOQSQ&cver=2.20210512.06.00&fallback_count=1&range=0-2771&rn=34&rbuf=0
[HTTP/1.1 403 Forbidden 235ms]

22:29:04.787 Error: Could not establish connection. Receiving end does not exist. 4 tabs.js:199:23
22:29:04.806 XHRGEThttps://r12---sn-bvvbax-cvne.googlevideo.com/videoplayback?expire=1621030536&ei=KKKeYPbyK4eZhgbxqaTYBQ&ip=2601%3A188%3Ac302%3Acbc0%3A0%3A0%3A0%3A1236&id=o-AOl1jCqpxEwWGgbTvfMOgIUXVf4ro0u2aqhXAHIjR9aI&itag=243&aitags=133%2C134%2C135%2C136%2C137%2C160%2C242%2C243%2C244%2C247%2C248%2C278&source=youtube&requiressl=yes&mh=n3&mm=31%2C29&mn=sn-bvvbax-cvne%2Csn-ab5l6nzd&ms=au%2Crdu&mv=m&mvi=12&pl=34&initcwndbps=1778750&vprv=1&mime=video%2Fwebm&ns=zg4iXPEq5BqKRfxq9gFOsGUF&gir=yes&clen=33571236&dur=766.160&lmt=1564032433790164&mt=1621008797&fvip=2&keepalive=yes&fexp=24001373%2C24007246&beids=23886210&c=WEB&txp=6316222&n=G7L2JZsf-_7lPQ&sparams=expire%2Cei%2Cip%2Cid%2Caitags%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRAIgPWu00OXGyy17HfDQF-AiOU-kgEfqjZf8-5W6ch9cZ6cCIBuZP1i9ziCgEpSTiQSLL5Xqi2IppXTO8GfIDa8daEhZ&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRAIgN4DFcyhMtLFD8BIV-AEPkVGC2vlu_amx68AnSzZhYHECIFDvETEEY8fgXuhmHaCtPtz5EsJV_N46jsCNzyKVVH4v&alr=yes&cpn=xsPuLxxHGAbFOQSQ&cver=2.20210512.06.00&range=0-2837&rn=35&rbuf=0
[HTTP/1.1 403 Forbidden 93ms]

22:29:04.818 Error: Could not establish connection. Receiving end does not exist. 2 tabs.js:199:23
`

**Comment by aaFn** [comment]
Hello @ttggzz , ok, none of these messages are related to BSP2. You have an add-on which has apparently a lot of errors (see tabs.js, background.js, context-menu.js .. 

Can't ell which ad-on this is, but when the errors happen and you are on the browser console, simply hover the mouse over the source file name (e.g. background.js), and you will see that you can click on it.
=> Click on those different errors, and FF will open the source code at the place where the error is.

Then, browsing the source code (maybe going at beginning), you could catch some keywords saying to you which add-on this is.
From what I see, this is an add-on related to tabs (maybe All Tabs Helper ?)

Apart from that, there is nothing abnormal showing in the browser console, so I have no clue about what is going on so far, sorry.

**Comment by ttggzz** [comment]
Thanks for taking a look! Haven't been able to get back to this yet, at the time briefly tried your instructions without being able to find tabs.js etc. If I recall couldn't find or see what to hover my mouse over. 

There was another freeze up and on a hunch I disabled 'all tabs helper', no change. I then disabled bsp2, re-enabled, and all was back to normal, same after I then re-enabled ATH. So if that works again could be a workaround for now, enable-disable being so easy in FF. All this was with no browser restart, 'on the fly', so to speak. 

I have done your latest update, but have not yet put any time on it in the browser.
thanks

**Comment by aaFn** [comment]
Ok, next time it happens, in case this is due to some internal never ending BSP2 loop which is logging nothing, here is what I propose:
- Expand FF full screen if not already
- Go to Tools -> add-ons
- Click on the little wheel at top right ("Tools for all add-ons") and click on "Debug add-ons"
- A new tab will appear with a list of extensions (you might have to click on the Extension twistie to show it)
- You will see BSP2 listed
- Click on the "Inspect" button inside it
- That will open a new tab, called the Toolbox, showing debug tools for add-ons, and focused on BSP2
- Click on the "Debugger" tab if not already the one appearing
- Then click on the "Pause" button at top right, or press F8
- If BSP2 was looping, then it will stop in code where it is
- If it doesn't stop, that means BSP2 was not running .. maybe wait a bit in case the signal takes time for FF to process, but if more than 30s of doing nothing, then that means BSP2 is probably not the culprit
  - In such a case, to confirm, click anywhere in the BSP2 tree, and a break should occur in code immediately or nearly
  - Press F8 again and close the tab, the problem must be in another add-on
- But if BSP2 stops on your first pause strike, that should confirm it is in a loop, and then continue as follows:
  - Look at the call stack on the right, expand it if needed
  - Also expand the Scopes section, and expand the first few top contexts (usually a function name, then "Block", but there can be more)
  - Take a full screen screenshot of the point in code where execution is stopped, including the full call stack, and the top of Scopes part
  - Paste it here
  - I will see from that if I can get enough information to find the cause

Thanks in advance, aaFn

**Comment by ttggzz** [comment]
Spectacularly detailed response, thanks for your time! I forgot to mention the most recent freeze-ups were occurring when trying out the 'advanced>collapse all in branch' feature, duplicated several times. 'Collapse all' worked, then search performed normally, then the freeze when clicking a search result.

Well I've got a pretty heavily loaded session going, intentionally let tabs stay open rather than unloading. Tried the 'collapse all' several times then search>click, including a second instance of bsp2, snappy/nary a hiccup. 

In the meantime I may print out your last response to have on hand in case of a future freeze, or perhaps keep that page open in my other active browser. Thanks again.

On another note, I too was looking for your donation button! Sorry I can't help you as to which venue to choose, but considering FF would be miserable for me without bsp2 your work is valuable. I am astonished to see bsp2 only has a few thousand users, I can't fathom what the millions of other FF users are doing with their libraries (if large). Tags maybe I guess, I remember examining that way back and was not attracted to the workflow.
